### ISSUE TITLE
Boost performance after massive add of devices

### ISSUE THREAD

**Issue by mp3merlin** [Cacti/cacti#1020]
As advised in my post on the Cacti forum ([https://forums.cacti.net/viewtopic.php?f=21&t=57899]), I open this issue.

I have a Cacti 1.1.16 test install with about 10000 devices which has been working fine for about 3 weeks.
All polling is done from a single remote poller (in the same LAN).
I decided to add about 2000 devices using the add_device.php CLI script. After the successful add, boost isn't running anymore.
I can still run a poller_boost.php --force to update the RRDs.

In my cacti.log (on the main server) I have some messages like this :

```
2017-08-30 17:50:09 - SYSTEM STATS: Time:8.2158 Method:spine Processes:4 Threads:50 Hosts:0 HostsPerProcess:0 DataSources:0 RRDsProcessed:0
2017-08-30 17:53:10 - ERROR PHP ERROR: Maximum execution time of 299 seconds exceeded in file: /var/www/cacti/lib/database.php on line: 374
2017-08-30 17:53:10 - CMDPHP PHP ERROR Backtrace: (CactiShutdownHandler)(/lib/functions.php: 4432 cacti_debug_backtrace)
2017-08-30 17:55:04 - SYSTEM STATS: Time:3.1443 Method:spine Processes:4 Threads:50 Hosts:0 HostsPerProcess:0 DataSources:0 RRDsProcessed:0
2017-08-30 17:57:32 - ERROR PHP ERROR: Maximum execution time of 299 seconds exceeded in file: /var/www/cacti/lib/snmpagent.php on line: 0
2017-08-30 17:57:32 - CMDPHP PHP ERROR Backtrace: (CactiShutdownHandler)(/lib/functions.php: 4432 cacti_debug_backtrace)
2017-08-30 18:00:12 - SYSTEM STATS: Time:10.4571 Method:spine Processes:4 Threads:50 Hosts:0 HostsPerProcess:0 DataSources:0 RRDsProcessed:0
2017-08-30 18:03:18 - ERROR PHP ERROR: Maximum execution time of 299 seconds exceeded in file: /var/www/cacti/lib/snmpagent.php on line: 823
2017-08-30 18:03:18 - CMDPHP PHP ERROR Backtrace: (CactiShutdownHandler)(/lib/functions.php: 4432 cacti_debug_backtrace)
2017-08-30 18:05:09 - SYSTEM STATS: Time:7.7355 Method:spine Processes:4 Threads:50 Hosts:0 HostsPerProcess:0 DataSources:0 RRDsProcessed:0
2017-08-30 18:06:53 - ERROR PHP ERROR: Maximum execution time of 299 seconds exceeded in file: /var/www/cacti/lib/snmpagent.php on line: 449
2017-08-30 18:06:53 - CMDPHP PHP ERROR Backtrace: (CactiShutdownHandler)(/lib/functions.php: 4432 cacti_debug_backtrace)
```
No problems on the poller :

```2017-08-30 18:16:51 - SYSTEM STATS: Time:107.8277 Method:spine Processes:4 Threads:50 Hosts:12387 HostsPerProcess:3097 DataSources:116298 RRDsProcessed:0```

I activated the mysql slow log.

```
# mysqldumpslow /var/log/mysql.slow.log

Reading mysql slow query log from /var/log/mysql.slow.log
Count: 294 Time=18.51s (5442s) Lock=0.00s (0s) Rows_sent=1.0 (294), Rows_examined=88606191.8 (26050220386), Rows_affected=0.0 (0), cactiuser[cactiuser]@localhost
SELECT count(*) FROM poller_output_boost_arch_1504509016

Count: 463 Time=5.85s (2709s) Lock=0.00s (0s) Rows_sent=1.0 (463), Rows_examined=27815985.8 (12878801408), Rows_affected=0.0 (0), cactiuser[cactiuser]@localhost
SELECT count(*) FROM poller_output_boost_arch_1504172552

Count: 958 Time=1.25s (1198s) Lock=0.00s (0s) Rows_sent=10.0 (9580), Rows_examined=1448706.5 (1387860838), Rows_affected=0.0 (0), cactiuser[cactiuser]@integsr-t241
SELECT DISTINCT field_name FROM host_snmp_cache AS hsc WHERE snmp_query_id= 'S'
```

On the forum, Osiris suggested to upgrade to 1.1.21 (latest at the time).
I did, but the situation is the same.

Is there anything I can do to diagnose this problem? My objective is to monitor about 25k devices.

**Comment by cigamit** [comment]
There are some changes that we are contemplating more in the 1.2 timeframe that should help you get to 25k devices using a the multiple poller architecture that will help you achieve that performance.  So, for now, you will just have to disable the snmpagent.

To do that in this cacti version, edit lib/snmpagent.php and goto approximately line 288 inside of the function snmpagent_poller_bottom and place a return; statement to return from this function immediately.  Let me know if this takes the heat off the data collection.

What I anticipate from a data collection perspective should help you, but I also believe that this snmpagent activity should be done in the background and not inline during the main data collection.  I will take this up with the author.

**Comment by mp3merlin** [comment]
Thanks. I'll try to do that as soon as I can and I'll keep you posted.

**Comment by mp3merlin** [comment]
I tried your suggestion on my lab install (the one that has the problem). After inserting a `return; ` in the `snmpagent_poller_bottom()` function, boost is working again and my RRDs are updated ;)

So now, I'm wondering what is the use of this snmpagent ?

Polling is pretty fast (mostly SNMP and a few TCL scripts) on a low end server (4CPUs only) : 

```
2017-10-19 14:07:11 - SYSTEM STATS: Time:127.7842 Method:spine Processes:4 Threads:50 Hosts:12387 HostsPerProcess:3097 DataSources:116873 RRDsProcessed:0
```

Boost update is taking around55s for 550k+ updates every 30 minutes on the central server (24CPUs but not really fast disks).

I'll try to add a few thousand devices to my lab platform to see how it goes.

**Comment by cigamit** [comment]
Good to know.  I've asked Browniebraun to address this.

**Comment by artiflo** [comment]
We have still this issue.

Somebody could be assigned ? :)

**Comment by phineasthecat** [comment]
I also ran into this problem, on cacti 1.1.38. The performance issues were so severe with our cacti server, it made the web interface almost unusable when the poller was running, and pages with lots of graphs took multiple minutes to load.

Put a `return`; in the first line of the `snmpagent_poller_bottom()` function and the performance problem went away... Is this fixed or planned to be fixed in 1.2?

**Comment by cigamit** [comment]
You can now disable the snmpagent from the poller tab.  I'll therefore close this.
